tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner_test.py

```diff
@@ -0,0 +1,343 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for the key functions in riglpruner library."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# import g3
+
+from absl.testing import parameterized
+import numpy as np
+import tensorflow as tf
+
+# TODO(b/139939526): move to public API.
+from tensorflow.python.keras import keras_parameterized
+from tensorflow_model_optimization.python.core.sparsity_tf2 import schedule
+from tensorflow_model_optimization.python.core.sparsity_tf2 import riglpruner as pruner
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruning_optimizer
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruning_config
+
+dtypes = tf.dtypes
+test = tf.test
+
+
+def make_update_schedule(fraction, begin, end, freq):
+  return schedule.ConstantSparsity(fraction, begin, end, freq)
+
+def sample_noise(x, mu=0, sigma=1.):
+  sample = tf.random.normal((), mean=mu,  stddev=sigma, dtype=tf.float64)
+  return sample
+
+def _dummy_gradient(x, dtype=tf.float32):
+  try:
+    base_type = x.dtype
+  except:
+    base_type = dtype
+  # must shuffle otherwise each gradient update will be the same and 
+  # connections are never regrown
+  grad = tf.random.shuffle(np.linspace(1., 100., 100))
+  return tf.Variable(grad, dtype=base_type)
+
+class RiglPruningTest(test.TestCase, parameterized.TestCase):
+
+  def setUp(self):
+    super(RiglPruningTest, self).setUp()
+    self.block_size = (1, 1)
+    self.block_pooling_type = "AVG"
+    self.target_sparsity = 0.5
+    self.initial_drop_fraction = 0.3 # i.e. 0.3 * target sparsity per layer, annealed by schedule below
+    self.constant_updater = schedule.ConstantSchedule(self.initial_drop_fraction, 0, 100, 1)
+    self.skip_updater = schedule.ConstantSchedule(self.initial_drop_fraction, 0, 100, 2)
+    self.updater = schedule.ConstantSchedule
+    self.grad = _dummy_gradient
+    self.seed = [0, 1]
+    self.noise_std = 1
+    self.reinit = False
+    self.stateless = True
+    self.grow_init_zeros = 'zeros'
+    self.grow_init_randn = 'random normal'
+    self.grow_init_randunif = 'random uniform'
+
+  def testGetGrowGrads(self):
+    mask = tf.ones((100))
+    grads = -1 * tf.linspace(1., 6., 6)
+    expected_grads = tf.linspace(1., 6., 6)
+
+    p = pruner.RiGLPruner(
+      update_schedule=self.skip_updater,
+      sparsity=self.target_sparsity,
+      block_size=self.block_size,
+      block_pooling_type=self.block_pooling_type,
+      stateless=self.stateless,
+      seed=self.seed,
+      noise_std=self.noise_std,
+      reinit=self.reinit,
+      grow_init=self.grow_init_zeros
+    )
+
+    grow_grads = p._get_grow_grads(mask, grads)
```

Comment:
This isn't a particularly useful test as written. It just checks the behavior of one minor utility method that just does an `abs` and we may choose to inline at any time.

A more useful test would be directly verifying what the pruner decides to grow. (whether via raw indices, or via a mask tensor that represents what should be grown)

Reply:
On top of that, even if the goal was to just check `abs`, this test would fail to distinguish between `abs(x)` and `-1 * x` because all the values are negative.

Reply:
I'm not sure why this conversation was marked resolved? These concerns all still apply. Can you check the actual grow mask and/or grow tensor instead?

Reply:
(or get rid of this test if you're checking the grow mask elsewhere)

Reply:
I wrote a specific test yesterday that manually calculated the grow indices for the mask

Reply:
Thanks Winnie! You can go ahead and delete this test then